fs-cache/src/cache.rs

```diff
@@ -0,0 +1,60 @@
+use crate::memory_limited_storage::MemoryLimitedStorage;
+use data_error::Result;
+use std::path::Path;
+
+pub struct Cache<K, V> {
+    storage: MemoryLimitedStorage<K, V>,
+}
+
+impl<K, V> Cache<K, V>
+where
+    K: Ord
+        + Clone
+        + serde::Serialize
+        + serde::de::DeserializeOwned
+        + std::fmt::Display
+        + std::hash::Hash
+        + std::str::FromStr,
+    V: Clone + serde::Serialize + serde::de::DeserializeOwned,
+{
+    pub fn new(
+        label: String,
+        path: &Path,
+        max_memory_bytes: usize,
+    ) -> Result<Self> {
+        log::debug!(
+            "{} cache initialized with {} bytes limit",
+            label,
+            max_memory_bytes
+        );
+        Ok(Self {
+            storage: MemoryLimitedStorage::new(label, path, max_memory_bytes)?,
+        })
+    }
+
+    pub fn get(&mut self, key: &K) -> Option<V> {
+        let result = self.storage.get(key);
+        log::debug!(
+            "{} cache: get key={} -> found={}",
+            self.storage.label(),
+            key,
+            result.is_some()
+        );
+        result
+    }
+
+    pub fn set(&mut self, key: K, value: V) -> Result<()> {
+        // Check if value already exists
```

Comment:
What happens if a value already exists? This behavior isn’t documented.

Reply:
Good catch. It can be indicator of redundant computation.

Since the cache uses `FolderStorage` under the hood, checking for key existence should be faster than retrieving the whole value.. Am I right that just checking the file existence is faster than reading the file?

I suggest the following:
- return error when the entry already exists
- add `keys` query to the API, that returns all known keys
- add `exists` predicate to the API

Query `keys` could be used by app developers to efficiently filter already known values, so they could skip not only writing to the disk, but entire redundant computation.